runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryPolicy.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries.impl
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.SdkBaseException
+import aws.smithy.kotlin.runtime.ServiceException
+import aws.smithy.kotlin.runtime.ServiceException.*
+import aws.smithy.kotlin.runtime.retries.RetryDirective
+import aws.smithy.kotlin.runtime.retries.RetryErrorType
+import aws.smithy.kotlin.runtime.retries.RetryPolicy
+
+/**
+ * A standard retry policy which attempts to derive information from the Smithy exception hierarchy.
+ */
+open class StandardRetryPolicy : RetryPolicy<Any?> {
+    /**
+     * Evaluate the given retry attempt.
+     */
+    override fun evaluate(result: Result<Any?>): RetryDirective = when {
+        result.isSuccess -> RetryDirective.TerminateAndSucceed
+        else -> evaluate(result.exceptionOrNull()!!)
+    }
+
+    private fun evaluate(ex: Throwable): RetryDirective =
+        (ex as? SdkBaseException)?.run(::evaluateBaseException)
+            ?: (ex as? ServiceException)?.run(::evaluateServiceException)
+            ?: (ex as? ClientException)?.run { RetryDirective.RetryError(RetryErrorType.ClientSide) }
+            ?: evaluateNonSdkException(ex)
+            ?: evaluateOtherExceptions(ex)
+            ?: RetryDirective.TerminateAndFail
+
+    private fun evaluateBaseException(ex: SdkBaseException): RetryDirective? = with(ex.sdkErrorMetadata) {
+        when {
+            isThrottling -> RetryDirective.RetryError(RetryErrorType.Throttling)
+            else -> null
+        }
+    }
+
+    private fun evaluateServiceException(ex: ServiceException): RetryDirective? = with(ex.sdkErrorMetadata) {
+        when {
+            isRetryable && errorType == ErrorType.Server -> RetryDirective.RetryError(RetryErrorType.ServerSide)
+            isRetryable && errorType == ErrorType.Client -> RetryDirective.RetryError(RetryErrorType.ClientSide)
+            else -> null
+        }
+    }
+
+    protected open fun evaluateOtherExceptions(ex: Throwable): RetryDirective? = null
+
+    private fun evaluateNonSdkException(ex: Throwable): RetryDirective? =
+        // TODO Write logic to find connection errors, timeouts, stream faults, etc.
```

Comment:
question

is this follow up work or blocked on something else?

Reply:
I'd hoped this very PR would spawn a discussion about how to fill this in. Do we have a sane way of detecting unmodeled exceptions for things like IO?

I'd imagine the `HttpClientEngine` implementations are what will be throwing most of these but the only engine-level exception I can find is [`HttpClientEngineClosedException`](https://github.com/awslabs/smithy-kotlin/blob/main/runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/engine/HttpClientEngine.kt#L63). Absent a common exception layer, it seems like the retry policy would have to know the individual exceptions thrown by implementations (which sounds too tightly coupled).

Do we need a more clearly defined set of exceptions thrown by engines? Are there non-engine sources of useful exceptions?

Reply:
It's true that we probably need to have more taxonomy for particular exceptions but I'd start with at least classifying transient errors off raw response codes (we can leave this to `aws-sdk-kotlin` if we like since the SEP for this behavior is AWS specific)